### PR TITLE
Add per-region SALINITY support for CO2STORE multi-PVT decks

### PR DESCRIPTION
**Pull request by dlmorenob**
## Summary                                                                                                                                                                         
                                                                                                                                                                                     
  The `SALINITY` keyword currently accepts only a single record, so all PVT regions in a CO2STORE model use the same brine salinity. This prevents accurate modelling of stacked     
  aquifer storage targets with different formation-water TDS, or formations with caprock/reservoir salinity contrasts.
                                                                                                                                                                                     
  This PR allows one `SALINITY` record per PVT region (size tied to `TABDIMS` item `NTPVT`, mirroring how `PVTW` is handled).                                                        
  
  ## Changes                                                                                                                                                                         
                                                                
  - `share/keywords/001_Eclipse300/S/SALINITY` — keyword size: `1` → `{"keyword": "TABDIMS", "item": "NTPVT"}`                                                                       
  - `Co2StoreConfig.hpp` — `double salt` → `std::vector<double> salt`; added `salinity(regionIdx)` and `numSalinityRegions()` accessors
  - `Co2StoreConfig.cpp` — parser loops over records; backward-compat `salinity()` returns `salt[0]`; out-of-range `regionIdx` broadcasts `salt[0]`                                  
  - `BrineCo2Pvt.cpp` — `salinity()` → `salinity(regionIdx)` in per-region init loop                                                                                                 
  - `Co2GasPvt.cpp` — same one-line fix for gas-phase PVT path                                                                                                                       
  - `tests/parser/EclipseStateTests.cpp` — added `TABDIMS /` to existing `SALTMFTest`                                                                                                
                                                                
  ## Backward compatibility                                                                                                                                                          
                                                                
  A deck with a single `SALINITY` record (the only form currently in use) works unchanged. `NTPVT` defaults to 1, `salinity()` (no args) returns `salt[0]`, and `salinity(regionIdx)`
   broadcasts `salt[0]` for out-of-range indices.
                                                                                                                                                                                     
  ## Testing                                                    

  - 23 unit-test assertions (single/multi-record, SALTMF, edge cases)                                                                                                                
  - Full ctest: 251/252 pass (1 pre-existing `python_tests` failure on clean master)
  - CO2STORE simulations: backward compat (−0.00% mass balance), extreme contrast (fresh vs near-halite: +194% dissolution difference confirms salting-out), uniform control (−1% difference confirms broadcast)                                                                                                                                                     
  - Cross-compiler: g++-13.3 and g++-14.2
                                                                                                                                                                                     
  ## Context                                                    
                                                                                                                                                                                     
  Second of two follow-ups requested by the reviewer of #5097. Independent of the BRINE/fugacity gating fix (to be filed separately).

### CODE REVIEW

**Comment by svenn-t** [comment]
Thanks for your work. 

Since `SALINITY` is a keyword from the commercial simulator, that expects a constant and not one value per PVT region, I'm against changing the definition of this keyword. I suggest making a new salinity keyword which sets salinity per pvt region and implement your changes using the new keyword instead.

**Comment by bska** [comment]
>  I suggest making a new salinity keyword which sets salinity per pvt region and implement your changes using the new keyword instead.

We might consider implementing the `SALINITR` keyword.  This doesn't precisely fit the request as that keyword has one value for each equation-of-state region (`EOSNUM`) instead of PVT region (`PVTNUM`), but maybe we could make it work per PVT region as well?

**Comment by svenn-t** [comment]
> > I suggest making a new salinity keyword which sets salinity per pvt region and implement your changes using the new keyword instead.
> 
> We might consider implementing the `SALINITR` keyword. This doesn't precisely fit the request as that keyword has one value for each equation-of-state region (`EOSNUM`) instead of PVT region (`PVTNUM`), but maybe we could make it work per PVT region as well?

Yes, that is a good suggestion. If an implementation based on `EOSNUM` does not give issues elsewhere (restarts?), I would do it that way.

**Comment by dlmorenob** [comment]
Thanks for the direction — SALINITR is OK. However, It's not in OPM's parser yet so this would be an addition, right?
                                                                                                                                                                                     
I'll check the implementation using  EOSNUM as you suggest, that said, I have not seen  CO2STORE decks that set  EOSNUM. Should SALINITR require EOSNUM in the deck, or default to PVTNUM when EOSNUM is absent?

**Comment by dlmorenob** [comment]
Following up on EOSNUM for SALINITR: the OPM Flow manual notes that EOSNUM is currently "parsed but its data ignored." Implementing SALINITR per EOSNUM would therefore require first wiring EOSNUM?
Is PVTNUM-based SALINITR still acceptable as an interim, or do you propose something else?

**Comment by svenn-t** [comment]
> Following up on EOSNUM for SALINITR: the OPM Flow manual notes that EOSNUM is currently "parsed but its data ignored." Implementing SALINITR per EOSNUM would therefore require first wiring EOSNUM? Is PVTNUM-based SALINITR still acceptable as an interim, or do you propose something else?

Yes, SALINITR might be difficult to implement due to the current state of EOSNUM. I propose then to make an entirely new keyword (not named SALINITR, since that is also defined in the commercial simulator) and implement PVTNUM-based salinity.

**Comment by dlmorenob** [comment]
Reworked per discussion: I introduce SALINITP under 900_OPM/S/, sized by NTPVT, prohibits SALINITY/SALTMF/BRINE. Existing scalar SALINITY/SALTMF unchanged. Per-region indexing  wired through BrineCo2Pvt, Co2GasPvt, and the existing CO2STORE-guarded init in BlackOilFluidSystem. New tests for per-region behavior, scalar broadcast, and parser rejection of conflicting keywords.

File: opm/input/eclipse/EclipseState/Co2StoreConfig.cpp [hunk 1]
```diff
@@ -127,14 +127,28 @@ namespace Opm {
...
+            salt.resize(numRecords);
+            for (std::size_t i = 0; i < numRecords; ++i) {
+                const auto& molality = keyword.getRecord(i).getItem("MOLALITY").get<double>(0);
+                salt[i] = 1.0 / (1.0 + 1.0 / (molality * MmNaCl));
+            }
+        }
+        else if (props_section.hasKeyword<ParserKeywords::SALINITY>()) {
             const auto& molality = deck["SALINITY"].back().getRecord(0).getItem("MOLALITY").get<double>(0);
-            salt = 1.0 / (1.0 + 1.0 / (molality * MmNaCl));
+            salt.resize(1);
```
Comment (svenn-t) [inline opm/input/eclipse/EclipseState/Co2StoreConfig.cpp:145]: Resize according to NTPVT from TABDIMS and fill vector with the one value from SALINITY.

File: opm/input/eclipse/EclipseState/Co2StoreConfig.cpp [hunk 2]
```diff
@@ -127,14 +127,28 @@ namespace Opm {
...
+            salt[0] = 1.0 / (1.0 + 1.0 / (molality * MmNaCl));
         }
         else if (props_section.hasKeyword<ParserKeywords::SALTMF>()) {
             const auto& mole_frac = deck["SALTMF"].back().getRecord(0).getItem("MOLE_FRACTION").get<double>(0);
-            salt = mole_frac * MmNaCl / (mole_frac * (MmNaCl - MmH2O) + MmH2O);
+            salt.resize(1);
```
Comment (svenn-t) [inline opm/input/eclipse/EclipseState/Co2StoreConfig.cpp:150]: Resize according to NTPVT from TABDIMS and fill vector with the one value from SALTMF.

File: opm/input/eclipse/EclipseState/Co2StoreConfig.cpp [hunk 3]
```diff
@@ -127,14 +127,28 @@ namespace Opm {
             initEzrokhiTable(deck, "VISCAQA", num_eos_res, cnames, viscaqa_tables);
         }
 
-        // SALINITY or SALTMF convert to mass fraction.
-        if (props_section.hasKeyword<ParserKeywords::SALINITY>()) {
+        // SALINITP / SALINITY / SALTMF convert to mass fraction.
+        // SALINITP (OPM-specific): per-PVT-region salinity, sized by NTPVT.
+        // SALINITY (commercial Eclipse): single scalar value, broadcast to all regions.
```
Comment (svenn-t) [inline opm/input/eclipse/EclipseState/Co2StoreConfig.cpp:132]: Remove "Eclipse"

File: opm/input/eclipse/EclipseState/Co2StoreConfig.cpp [hunk 4]
```diff
@@ -155,7 +169,20 @@ namespace Opm {
     }
 
     double Co2StoreConfig::salinity() const {
-        return salt;
+        return salt.empty() ? 0.0 : salt[0];
+    }
+
```
Comment (svenn-t) [inline opm/input/eclipse/EclipseState/Co2StoreConfig.cpp:174]: Remove `salinity()` function without any input. We will only use the version with `regionIdx` input.

File: opm/input/eclipse/EclipseState/Co2StoreConfig.cpp [hunk 5]
```diff
@@ -155,7 +169,20 @@ namespace Opm {
...
+    double Co2StoreConfig::salinity(std::size_t regionIdx) const {
+        if (salt.empty())
+            return 0.0;
+        // Broadcast single-record (SALINITY/SALTMF) value to all regions.
+        if (regionIdx >= salt.size())
```
Comment (svenn-t) [inline opm/input/eclipse/EclipseState/Co2StoreConfig.cpp:179]: Remove broadcasting. SALINITY and SALTMF should be the same size as NTPVT.

File: opm/input/eclipse/EclipseState/Co2StoreConfig.cpp [hunk 6]
```diff
@@ -155,7 +169,20 @@ namespace Opm {
...
+            return salt[0];
+        return salt[regionIdx];
+    }
+
+    std::size_t Co2StoreConfig::numSalinityRegions() const {
```
Comment (svenn-t) [inline opm/input/eclipse/EclipseState/Co2StoreConfig.cpp:184]: Remove. We can use NTPVT information if necessary.

**Review by svenn-t — CHANGES_REQUESTED** [review]
I have some minor comments, mostly related to enforcing the `salt` vector to be NTPVT size regardless of keyword.
